finish implementation of upgradable stdlibs

vtjnash opened the pull request:
Fixes #53983

~~Unfortunately, TOML has a design problem with the way it is used by Pkg (because of the way they use Dates without depending on it correctly, they want to assume things that are false now), so this does not yet work correctly for Pkg. But maybe if we change some of the uses of loaded_packages to explicit_loaded_modules in loading, we can try to make the way those three packages interact to still kind of mostly work.~~ (Fixed in Pkg now and removed from TOML in https://github.com/JuliaLang/julia/pull/54755)

~~Also currently one test is broken, as we need some more code rearrangement to make it legal again for an `__init__` method to import code.~~

vtjnash committed:
upgradable stdlibs

Fixes #53983

This breaks one test, which relied on it being undefined behavior to
rely on the execution of `__init__` having occurred before `using` is
allowed to return. That usability issue is now (partially) eliminated.

JeffBezanson commented on base/loading.jl:
```diff
@@ -2289,14 +2300,19 @@ end
-const explicit_loaded_modules = Dict{PkgId,Module}()
+const explicit_loaded_modules = Dict{PkgId,Module}() # Emptied on Julia start
+const loaded_modules = Dict{PkgId,Module}() # available to be explicitly loaded
+const loaded_precompiles = Dict{Pair{PkgId,UInt128},Module}() # extended (complete) list of modules, available to be loaded
```
Can we be clearer about the differences between these three dicts? Is the first one modules loaded during this session? Is the second one currently visible modules, and the third one all modules? I'm not sure.

matthias314 commented:
I'm sorry, but this doesn't fully fix #53983. There are still problems when using `startup.jl`. The setting is as follows: no `~/.julia` directory, `JULIA_DEPOT_PATH=/tmp/julia-depot` containing the file `/tmp/julia-depot/config/startup.jl` with contents
```
using InteractiveUtils
```
Then:
```
/usr/local/git/julia$ ./julia
[ Info: Precompiling InteractiveUtils [b77e0a4c-d291-57a0-90e8-8db25a27a240] 
[ Info: Precompiling REPL [3fa0cd96-eef1-5676-8a61-b3b8758bbffb] 
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.12.0-DEV.766 (2024-06-21)
 _/ |\__'_|_|_|\__'_|  |  Commit 9d8ecaa899 (0 days old master)
|__/                   |

julia>

[ now I press the key "]" ]

(@v1.12) pkg> Unhandled Task ERROR: ArgumentError: Package REPLExt [e5eb5ef1-03cf-53a7-ae1d-5a66b08e832b] is required but does not seem to be installed:
 - Run `Pkg.instantiate()` to install all recorded dependencies.
```
Starting Julia with `--startup-file=no` works fine. Unsetting `JULIA_DEPOT_PATH` and moving `startup.jl` to `~/.julia/config` instead also works fine.
```
Julia Version 1.12.0-DEV.766
Commit 9d8ecaa899 (2024-06-21 17:00 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 4 × Intel(R) Core(TM) i3-10110U CPU @ 2.10GHz
  WORD_SIZE: 64
  LLVM: libLLVM-17.0.6 (ORCJIT, skylake)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
Environment:
  JULIA_DEPOT_PATH = /tmp/julia-depot
```

vtjnash commented:
Much appreciated. I had fixed this earlier in the PR history, then forgot why I had the code to handle this and removed it again. Now it is clear again why I needed to handle this case with that example.

Keno commented:
I haven't fully bisected, but it seems like this is causing the following:
```
ERROR: InitError: KeyError: key Base.PkgId(Base.UUID("872c559c-99b0-510c-b3b7-b6c96a88d5cd"), "NNlib") not found
Stacktrace:
  [1] getindex
    @ ./dict.jl:477 [inlined]
  [2] macro expansion
    @ ./lock.jl:273 [inlined]
  [3] root_module(key::Base.PkgId)
    @ Base ./loading.jl:2358
  [4] parse_pkg_files(id::Base.PkgId)
    @ Revise ~/.julia/dev/Revise/src/loading.jl:38
  [5] watch_package(id::Base.PkgId)
    @ Revise ~/.julia/dev/Revise/src/pkgs.jl:349
  [6] add_require(sourcefile::String, modcaller::Module, idmod::String, modname::String, expr::Expr)
    @ Revise ~/.julia/dev/Revise/src/pkgs.jl:188
  [7] top-level scope
    @ ~/.julia/packages/Requires/Z8rfN/src/require.jl:70
  [8] (::NNlib.var"#599#603")()
    @ NNlib ~/.julia/packages/Requires/Z8rfN/src/require.jl:106
  [9] top-level scope
    @ ~/.julia/packages/Requires/Z8rfN/src/require.jl:20
 [10] macro expansion
    @ ~/.julia/packages/Requires/Z8rfN/src/require.jl:98 [inlined]
 [11] (::NNlib.var"#598#602")()
    @ NNlib ~/.julia/packages/Requires/Z8rfN/src/init.jl:11
 [12] __init__()
    @ NNlib ~/.julia/packages/Requires/Z8rfN/src/init.jl:18
 [13] run_module_init(mod::Module, i::Int64)
    @ Base ./loading.jl:1305
 [14] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
    @ Base ./loading.jl:1293
```

KristofferC commented:
Should we use https://github.com/JuliaLang/julia/pull/53849 for this then?

Keno commented:
> Should we use #53849 for this then?

That helps for pre-merge versioning. Now this can just use a regular VERSION check if intended or a fix if not.

KristofferC commented:
So this PR should have added an entry?

Keno commented:
If the break here was intentional, then what should have happened would be:
1. This PR adds an entry
2. We make the package work both ways by querying the entry
3. We merge/tag the package
4. We merge the PR here

KristofferC commented:
Any change to an internal is possibly breaking. And so far it seems no new entry has been added so I guess no internals have been broken? But Revise doesn't work now. So what should be changed in the workflow to make this thing actually do something?

KristofferC commented:
I am worried backporting this will cause https://github.com/JuliaLang/julia/issues/54940 on 1.11 but will have to check.
